### ISSUE TITLE
[Module Proposal]: `avm/utl/general/create-valid-az-resource-name`

### ISSUE THREAD

**Issue by SvenAelterman** [Azure/Azure-Verified-Modules#1156]
### Check for previous/existing GitHub issues/module proposals

- [X] I have checked for previous/existing GitHub issues/module proposals.

### Check this module doesn't already exist in the module indexes

- [X] I have checked for that this module doesn't already exist in the module indexes.

### Bicep or Terraform?

Bicep

### Module Classification?

I'm not sure, please advise

### Module Name

avm/utl/general/create-valid-az-resource-name

### Module Details

A module to create a valid Azure resource name given inputs such as naming convention structure, resource type, etc. It encodes the knowledge around character and length validity of resource names based on the resource type and it will generate a string which will be a valid (though not necessarily globally unique) resource name in Azure.

See an implementation at https://github.com/SvenAelterman/BicepTemplate/blob/main/common-modules/createValidAzResourceName.bicep

### Do you want to be the owner of this module?

Yes

### Module Owner's GitHub Username (handle)

SvenAelterman

**Comment by matebarabas** [comment]
@author, thank you for proposing this module and thank you for your patience on triaging it! Many of us on the AVM core team have been OOF in the last couple of weeks, but we're back now! 😊

As this is categorized as a utility module, its approval will require a more thorough assessment. The AVM core team will get back to you soon. Thank you for your patience!

**Comment by SvenAelterman** [comment]
Just checking on this. There seems to be a similar (at least based on name) proposal for TF. It might make sense to ensure that these modules will provide the same functionality across both TF and Bicep.

**Comment by matebarabas** [comment]
@SvenAelterman, I'll ping you internally.

**Comment by Dylan-Prins** [comment]
any updates on this? would love to have this. if not not the resource name, maybe we could look at region shortnames and abbreviations?

**Comment by SvenAelterman** [comment]
@Dylan-Prins I have a module that would perform this available in my own repo, which you're welcome to use. It doesn't follow AVM guidelines though.
https://github.com/SvenAelterman/BicepTemplate/blob/main/common-modules/createValidAzResourceName.bicep

I would be happy to obtain your feedback about it. Feel free to start a discussion in that repo.

**Comment by Dylan-Prins** [comment]
Hi Sven, Thanks for reaching out. I have something locally as well, but I want something i can use over multiple repo's.
Thats why I started looking on what AVM has available

**Comment by SvenAelterman** [comment]
@Dylan-Prins Understood. Are you familiar with the ability to set up your own Bicep module repository? This would be similar to the public repo in how you use it.

**Comment by Dylan-Prins** [comment]
Yes I am, but I am working for a MSP with a lot of customers. We move our bicep code to our customers, but don't want to use our ACR to provide a naming convention.

but thanks for the advise!
